runtime/sync_posix.h

```diff
@@ -0,0 +1,140 @@
+/**************************************************************************/
+/*                                                                        */
+/*                                 OCaml                                  */
+/*                                                                        */
+/*          Xavier Leroy and Damien Doligez, INRIA Rocquencourt           */
+/*                                                                        */
+/*   Copyright 1996 Institut National de Recherche en Informatique et     */
+/*     en Automatique.                                                    */
+/*                                                                        */
+/*   All rights reserved.  This file is distributed under the terms of    */
+/*   the GNU Lesser General Public License version 2.1, with the          */
+/*   special exception on linking described in the file LICENSE.          */
+/*                                                                        */
+/**************************************************************************/
+
+/* POSIX thread implementation of the user facing Mutex and Condition */
+/* To be included in runtime/sync.c */
+
+#include <errno.h>
+#include <pthread.h>
+#include <string.h>
+
+typedef int sync_retcode;
+
+/* Mutexes */
+
+typedef pthread_mutex_t * sync_mutex;
+
+#define Mutex_val(v) (* ((sync_mutex *) Data_custom_val(v)))
+
+Caml_inline int sync_mutex_create(sync_mutex * res)
+{
+  int rc;
+  pthread_mutexattr_t attr;
+  sync_mutex m;
+
+  rc = pthread_mutexattr_init(&attr);
+  if (rc != 0) goto error1;
+  rc = pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_ERRORCHECK);
+  if (rc != 0) goto error2;
+  m = caml_stat_alloc_noexc(sizeof(pthread_mutex_t));
```

Comment:
Probably orthogonal to this PR, but can't the memory allocation be moved up to the top and just use `caml_stat_alloc` to reduce the various `goto error` paths later?  caml_stat_alloc will raise Out_of_memory immediately, just as sync_error does later on anyway.

Reply:
This code is taken verbatim from otherlibs/systhreads/st_posix.h in OCaml 4, so I'd rather not change it in this PR, just trying to find a good place for this code.